standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java

```diff
@@ -20,71 +20,646 @@
+ * method determineDatabaseProduct is invoked.
+ * Tests that need to create multiple instances can use the reset method
+ * */
+public class DatabaseProduct implements Configurable {
+  static final private Logger LOG = LoggerFactory.getLogger(DatabaseProduct.class.getName());
+
+  private static enum DbType {DERBY, MYSQL, POSTGRES, ORACLE, SQLSERVER, EXTERNAL, OTHER};
+  public DbType dbType;
+  
+  // Singleton instance
+  private static DatabaseProduct theDatabaseProduct;
+
+  static {
+    final Configuration conf = MetastoreConf.newMetastoreConf();
+    // Check if we are using an external database product
+    boolean isExternal = MetastoreConf.getBoolVar(conf, ConfVars.USE_CUSTOM_RDBMS);
+
+    if (isExternal) {
+      // The DatabaseProduct will be created by instantiating an external class via
+      // reflection. The external class can override any method in the current class
+      String className = MetastoreConf.getVar(conf, ConfVars.CUSTOM_RDBMS_CLASSNAME);
+      
+      if (className != null) {
+        try {
+          theDatabaseProduct = (DatabaseProduct)
+              ReflectionUtils.newInstance(Class.forName(className), conf);
+
+          theDatabaseProduct.dbType = DbType.EXTERNAL;
+        }catch (Exception e) {
+          LOG.warn("Unable to instantiate custom database product. Reverting to default", e);
+        }
+      }
+      else {
+        LOG.warn("metastore.use.custom.database.product was set, " +
+                 "but metastore.custom.database.product.classname was not. Reverting to default");
+      }
+    }
+  }
+
+  /**
+   * Private constructor for singleton class
+   * @param id
+   */
+  private DatabaseProduct() {}
+  
+  public static final String DERBY_NAME = "derby";
+  public static final String SQL_SERVER_NAME = "microsoft sql server";
+  public static final String MYSQL_NAME = "mysql";
+  public static final String POSTGRESQL_NAME = "postgresql";
+  public static final String ORACLE_NAME = "oracle";
+  public static final String OTHER_NAME = "other";
+  
   /**
    * Determine the database product type
    * @param productName string to defer database connection
    * @return database product type
    */
-  public static DatabaseProduct determineDatabaseProduct(String productName) throws SQLException {
+  public static DatabaseProduct determineDatabaseProduct(String productName) {
+    DbType dbt;
+
     if (productName == null) {
-      return OTHER;
+      productName = OTHER_NAME;
     }
+
     productName = productName.toLowerCase();
-    if (productName.contains("derby")) {
-      return DERBY;
-    } else if (productName.contains("microsoft sql server")) {
-      return SQLSERVER;
-    } else if (productName.contains("mysql")) {
-      return MYSQL;
-    } else if (productName.contains("oracle")) {
-      return ORACLE;
-    } else if (productName.contains("postgresql")) {
-      return POSTGRES;
+
+    if (productName.contains(DERBY_NAME)) {
```

Comment:
Can we refactor these lines (108-119) into a separate private method which can be used like below at line 100?

if (theDatabaseProduct != null) {
  Preconditions.checkState(this.dbType == getDbType(productName));
  return theDatabaseProduct;
}